Handle *_between params in Reports controllers

jeffstieler opened the pull request:
Fixes #1239

Adds a utility method to `WC_Admin_Reports_Interval` to normalize numeric range`_between` parameters into `_min` and `_max`. It's only used by the Customers Report endpoint now, but can be used for all reports that make use of the advanced numeric filter.

### Accessibility

N/A - this is purely in the REST API.

### Detailed test instructions:

- Verify that the `phpunit` tests pass
- Use Postman or another tool to make some test requests to `/wc/v3/reports/customers`
- Test these params: `orders_count_between`, `total_spend_between`, `avg_order_value_between`

jeffstieler committed:
Add utility method to normalize numeric “between” parameters in REST API requests.

jeffstieler committed:
Handle *_between parameters in Customers Report endpoint controller.

joshuatf commented on includes/api/class-wc-admin-rest-reports-customers-controller.php:
```diff
@@ -402,7 +404,7 @@ public function get_collection_params() {
 		$params['total_spend_between']     = array(
 			'description'       => __( 'Limit response to objects with a total order spend between two given numbers.', 'wc-admin' ),
-			'type'              => 'array',
+			'type'              => 'string',
```
Curious if there's a reason we're opting to use strings here and parse inside `normalize_between_params` instead of typing it as an array.  Is this because we don't know if they've filled in both between fields?

```php
		$params['total_spend_between']     = array(
			'description'       => __( 'Limit response to objects with a total order spend between two given numbers.', 'wc-admin' ),
			'type'              => 'array',
			'items'             => array(
				'type' => 'float',
			),
			'validate_callback' => 'rest_validate_request_arg',
		);
```

jeffstieler replied:
I went with string parsing here because the in its current form, the advanced number filter produces a comma delimited value, not an array. I'll take a look and see how difficult it might be to change that (or parse on the JS side) since I agree that its cleaner to accept an array at the endpoint.

jeffstieler replied:
Switched back to an array in https://github.com/woocommerce/wc-admin/pull/1260/commits/ad9cbf4f6e3264e2d06fbea2f80778fc35eb4e35.

jeffstieler replied:
I'll open a separate PR to switch the numeric filter to generate an array.

joshuatf commented on includes/class-wc-admin-reports-interval.php:
```diff
@@ -474,4 +474,38 @@ public static function intervals_missing( $expected_interval_count, $db_records,
+			$max_param = $param_name . '_max';
+			$range     = explode( ',', $request[ $param_name . '_between' ] );
+
+			if ( isset( $range[0] ) && is_numeric( $range[0] ) ) {
```
Same question here.  I would expect a count of 2 to be given or assume this would be malformed and expected the client side to validate having both fields.  But maybe leaving off the front-end validation (as it is now) is easier for users.

jeffstieler replied:
In https://github.com/woocommerce/wc-admin/pull/1260/commits/ad9cbf4f6e3264e2d06fbea2f80778fc35eb4e35, I added validation and more normalization logic - the lower numeric value is always set to `_min`.

joshuatf reviewed (commented):
Looking really good and testing well!  This is a really great approach to normalizing between params 🎉 

One question about whether or not we should enforce the lower and upper values for this param to be valid, but otherwise this looks ready to merge.

jeffstieler committed:
Expect array value for _between params.

Add param validation method and normalize to min/max based by comparing values.

timmyc commented on tests/api/reports-interval.php:
```diff
@@ -796,4 +796,45 @@ public function test_next_year_start() {
 		}
 	}
 
+	/**
```
Thanks for the tests Jeff 🌟

timmyc reviewed (approved):
Once Jeff informed me I was testing the wrong endpoint, all is working well for me with this PR. Thanks for the tests too.
